### PR TITLE
Make it hard to accidentally enable an experimental feature flag

### PR DESCRIPTION
**Pull request by mkuratczyk**
Currently it's easy to enable an experimental feature flag like `khepri_db` by accident. Given the operation is irreversible, this leads to serious consequences when done in a production environment.

This PR implements the following changes:
* Management UI
  * Split the feature flags page into two tables - first with all but experimental feature flags, the second with only the experimental flags (and a warning)
  * To enable an experimental feature flag, a checkbox needs to be toggled first
* CLI
  * `rabbitmqctl enable_feature_flag khepri_db` won't work unless a new flag `--force` is provided

Screenshot 2024-08-14 at 12 19 57

### CODE REVIEW

**Comment by dumbbell** [comment]
What about naming the CLI flag `--experimental` to make sure the user knows it’s is an experimental feature flag?

Also, I feel that `--force` is too generic and could mean many things. I could interpret it as "I want this feature flag to be enabled even if some nodes are not running in the cluster" (which is denied today for instance).

**Comment by michaelklishin** [comment]
I agree with @dumbbell and would use `--experimental` instead of `--force`.

**Comment by michaelklishin** [comment]
The forced push was a rebase.

**Comment by mkuratczyk** [comment]
I've renamed the flag and slightly changed the error message (+rebase)

File: deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex [hunk 1]
```diff
@@ -29,32 +32,45 @@ defmodule RabbitMQ.CLI.Ctl.Commands.EnableFeatureFlagCommand do
...
-      # Server does not support feature flags, consider none are available.
-      # See rabbitmq/rabbitmq-cli#344 for context. MK.
-      {:badrpc, {:EXIT, {:undef, _}}} -> {:error, :unsupported}
-      {:badrpc, _} = err -> err
-      other -> other
+  def run([feature_flag], %{node: node_name, experimental: experimental}) do
+    case {experimental, :rabbit_misc.rpc_call(node_name, :rabbit_feature_flags, :get_stability, [
+      String.to_atom(feature_flag)
+    ])} do
+          {_, {:badrpc, {:EXIT, {:undef, _}}}} -> {:error, :unsupported}
```
Comment (dumbbell) [inline deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex:39]: I think you can drop the `undef` line: a node with this patch can’t talk to a 3.7 node anyway.

File: deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex [hunk 2]
```diff
@@ -7,15 +7,18 @@
 defmodule RabbitMQ.CLI.Ctl.Commands.EnableFeatureFlagCommand do
   @behaviour RabbitMQ.CLI.CommandBehaviour
 
-  def merge_defaults(args, opts), do: {args, opts}
+  def switches(), do: [experimental: :boolean]
+  def aliases(), do: [f: :experimental]
```
Comment (dumbbell) [inline deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex:11]: I have a hard time reading Elixir, but does this means that `-f` is an alternative to `--experimental`?

File: deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex [hunk 3]
```diff
@@ -29,32 +32,45 @@ defmodule RabbitMQ.CLI.Ctl.Commands.EnableFeatureFlagCommand do
...
 
   def usage_additional() do
     [
       [
         "<feature_flag>",
         "name of the feature flag to enable, or \"all\" to enable all supported flags"
+      ],
+      [
+        "--experimental",
+        "required to enable experimental feature flags (make sure you understand the risks!))"
```
Comment (dumbbell) [inline deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex:73]: There are two closing brackets in this sentence.

File: deps/rabbitmq_management/priv/www/js/tmpl/feature-flags.ejs [hunk 4]
```diff
@@ -76,3 +79,74 @@
...
 </div>
+
+
+
+<div class="section">
+  <h2>Experimental Feature Flags</h2>
+  <div class="hider">
+<% if (feature_flags.length > 0) { %>
+<p class="warning">
+  Feature flags listed below are experimental. They should not be enabled in a produciton deployment.
```
Comment (dumbbell) [inline deps/rabbitmq_management/priv/www/js/tmpl/feature-flags.ejs:90]: There is a typo in "production".

**Review by dumbbell — CHANGES_REQUESTED** [review]

**Comment by mkuratczyk** [comment]
Thanks for the feedback, I've pushed another commit fixing all the issues you pointed out. One question that came to mind is what should happen when `rabbitmqctl enable_feature_flag --experimental all` is invoked. Currently it still won't enable experimental flags and while intuitively, it should, given the goal of this PR is to prevent accidental enablement of experimental features, I feel like ignoring `--experimental` when `all` is used is fine. I can easily be persuaded though. Thoughts?

**Comment by dumbbell** [comment]
Instead of ignoring, I think we should either do what it says (enable everithing including experimental feature flags), or reject the command with an error.

As you said, given the goal, I would problably reject the command because the risks will be very specific to each experimental feature  flag.

File: deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex [hunk 5]
```diff
@@ -7,54 +7,63 @@
...
-      other -> other
+  def run([feature_flag], %{node: node_name, experimental: experimental}) do
+    case {experimental, :rabbit_misc.rpc_call(node_name, :rabbit_feature_flags, :get_stability, [
+      String.to_atom(feature_flag)
+    ])} do
+          {_, {:badrpc, _} = err} -> err
+          {false, :experimental} ->
+              IO.puts("Feature flag #{feature_flag} is experimental. If you understand the risk, use --experimental to enable it.")
+          _ ->
+                case :rabbit_misc.rpc_call(node_name, :rabbit_feature_flags, :enable, [
```
Comment (dumbbell) [inline deps/rabbitmq_cli/lib/rabbitmq/cli/ctl/commands/enable_feature_flag_command.ex:40]: As you removed the existing `undef` case clause, could you please put a note about why in the commit message? Because this is unrelated to the patch otherwise.

**Comment by mkuratczyk** [comment]
`--experimental` is now rejected when `all` flags are to be enabled. Commit message improved

**Comment by michaelklishin** [comment]
The forced push was a rebase.

**Review by michaelklishin — APPROVED** [review]
